File: min_ct_src.c
/* min_ct_src.c - Connected Topics minimal example publisher
 *
 * See https://github.com/UltraMessaging/lbmct
 *
 * Copyright (c) 2005-2018 Informatica Corporation. All Rights Reserved.
 * Permission is granted to licensees to use or alter this software for
 * any purpose, including commercial applications, according to the terms
 * laid out in the Software License Agreement.
 *
 * This source code example is provided by Informatica for educational
 * and evaluation purposes only.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INFORMATICA DISCLAIMS ALL WARRANTIES
 * EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION, ANY IMPLIED WARRANTIES OF
 * NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 * INFORMATICA DOES NOT WARRANT THAT USE OF THE SOFTWARE WILL BE UNINTERRUPTED
 * OR ERROR-FREE.  INFORMATICA SHALL NOT, UNDER ANY CIRCUMSTANCES, BE LIABLE TO  * LICENSEE FOR LOST PROFITS, CONSEQUENTIAL, INCIDENTAL, SPECIAL OR INDIRECT
 * DAMAGES ARISING OUT OF OR RELATED TO THIS AGREEMENT OR THE TRANSACTIONS
 * CONTEMPLATED HEREUNDER, EVEN IF INFORMATICA HAS BEEN APPRISED OF THE
 * LIKELIHOOD OF SUCH DAMAGES.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#ifdef _WIN32
    #include <winsock2.h>
#else
    #include <unistd.h>
    #include <netinet/in.h>
    #include <arpa/inet.h>
#endif
#include <lbm/lbm.h>
#include "lbmct.h"
#include "lbmct_private.h"
#include "prt.h"


/* State that app can be in. */
#define PRE_CREATE 0
#define CREATE_CALLED 1
#define DELETE_CALLED 2

int app_state = PRE_CREATE;


/* Very simplistic error handling macro.  Pass in return value of LBM function.
 * Prints error and aborts (core dump). */
#define LBM_ERR(_e) do {\
  if ((_e) != LBM_OK) {\
    fprintf(stderr, "LBM_ERR failed at %s:%d (%s)\n",\
       BASENAME(__FILE__), __LINE__, lbm_errmsg());\
    fflush(stderr);\
    abort();\
  }\
} while (0)


void *minsrc_src_conn_create_cb(lbmct_src_conn_t *src_conn,
  lbmct_peer_info_t *peer_info, void *src_clientd)
{
  char *s;

  /* As an example, capture the receiver's metadata and use it as the
   * connection state.
   */
  if (peer_info->flags & LBMCT_PEER_INFO_FLAGS_SRC_METADATA) {
    s = strdup(peer_info->rcv_metadata);
  }
  else {
    s = strdup("no metadata");
  }
  printf("conn create: peer='%s'\n", s);

  app_state = CREATE_CALLED;

  return s;
}  /* minsrc_src_conn_create_cb */


void minsrc_src_conn_delete_cb(lbmct_src_conn_t *src_conn,
  lbmct_peer_info_t *peer_info, void *src_clientd, void *conn_clientd)
{
  printf("conn delete: peer='%s'\n", conn_clientd);

  app_state = DELETE_CALLED;

  free(conn_clientd);
} /* minsrc_src_conn_delete_cb */


int main(int argc, char **argv) {
  lbm_context_attr_t *ctx_attr;
  lbm_context_t *ctx;
  lbmct_t *ct;
  lbmct_src_t *ct_src;
  lbm_src_t *um_src;
  int err;

  /* If config file supplied, read it. */
  if (argc > 1) {
    err = lbm_config(argv[1]);
    LBM_ERR(err);
  }

  /* Create context, making sure "response_tcp_nodelay" is set.
   */
  err = lbm_context_attr_create(&ctx_attr);
  LBM_ERR(err);
  err = lbm_context_attr_str_setopt(ctx_attr,
    "response_tcp_nodelay", "1");
  LBM_ERR(err);

  err = lbm_context_create(&ctx, ctx_attr, NULL, NULL);
  LBM_ERR(err);

  err = lbm_context_attr_delete(ctx_attr);
  LBM_ERR(err);

  /* Create CT object. */
  err = lbmct_create(&ct, ctx, NULL, "MinSrcMetadata", 15);
  LBM_ERR(err);


  /* Create connected source and wait connected. */
  err = lbmct_src_create(&ct_src, ct, "MinTopic", NULL, NULL,
    minsrc_src_conn_create_cb, minsrc_src_conn_delete_cb, NULL);
  LBM_ERR(err);

  um_src = lbmct_src_get_um_src(ct_src);

  while (app_state < CREATE_CALLED) {
    SLEEP_MSEC(100);
  }


  /* Send a message. */
  err = lbm_src_send(um_src, "Hello!", 7, LBM_MSG_FLUSH);
  LBM_ERR(err);

  /* Clean up. */
  err = lbmct_src_delete(ct_src);
  LBM_ERR(err);

  /* Wait for graceful close. */
  while (app_state < DELETE_CALLED) {
    SLEEP_MSEC(100);
  }

  err = lbmct_delete(ct);
  LBM_ERR(err);

  err = lbm_context_delete(ctx);
  LBM_ERR(err);

  return 0;
}  /* main */

File: java/MinCtSrc.java
/*
"MinCtSrc.java - Minimal Connected Topics source program.
 * See https://github.com/UltraMessaging/lbmct for code and documentation.
 *
 * Copyright (c) 2018-2019 Informatica Corporation. All Rights Reserved.
 * Permission is granted to licensees to use or alter this software for
 * any purpose, including commercial applications, according to the terms
 * laid out in the Software License Agreement.
 -
 - This source code example is provided by Informatica for educational
 - and evaluation purposes only.
 -
 - THE SOFTWARE IS PROVIDED "AS IS" AND INFORMATICA DISCLAIMS ALL WARRANTIES
 - EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION, ANY IMPLIED WARRANTIES OF
 - NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 - INFORMATICA DOES NOT WARRANT THAT USE OF THE SOFTWARE WILL BE UNINTERRUPTED
 - OR ERROR-FREE.  INFORMATICA SHALL NOT, UNDER ANY CIRCUMSTANCES, BE
 - LIABLE TO LICENSEE FOR LOST PROFITS, CONSEQUENTIAL, INCIDENTAL, SPECIAL OR
 - INDIRECT DAMAGES ARISING OUT OF OR RELATED TO THIS AGREEMENT OR THE
 - TRANSACTIONS CONTEMPLATED HEREUNDER, EVEN IF INFORMATICA HAS BEEN APPRISED
 - OF THE LIKELIHOOD OF SUCH DAMAGES.
 */

import java.util.*;
import java.nio.*;
import java.nio.charset.StandardCharsets;
import com.latencybusters.lbm.*;
import com.latencybusters.lbmct.*;

class MinCtSrc {
  // Have "main" create an object instance for this program and run it.
  private void run(String[] args) throws Exception {
    // Create context, making sure "response_tcp_nodelay" is set (speed up).
    LBMContextAttributes myCtxAttr = new LBMContextAttributes();
    myCtxAttr.setValue("response_tcp_nodelay", "1");
    LBMContext myCtx = new LBMContext(myCtxAttr);
    myCtxAttr.dispose();  // Recommended UM best practice.

    // Create a ByteBuffer of application metadata.
    ByteBuffer myMetadata = ByteBuffer.allocate(100);
    myMetadata.put("MySrcMeta".getBytes());  // Use a simple string as metadata.
    myMetadata.flip();    // Done writing metadata, present it to CT for reading.

    // Create CT object with a minimal configuration.
    LbmCtConfig myConfig = new LbmCtConfig();
    LbmCt myCt = new LbmCt();  // This just creates the object.
    myCt.start(myCtx, myConfig, myMetadata);  // Initialize the CT object.

    // Prepare to create the connected source.
    LBMSourceAttributes mySrcAttr = null;  // Or new LBMSourceAttributes();
    // Instantiate your source event, connect, disconnect callback objects.
    LBMSourceEventCallback myUmSrcEventCb = null;  // Not used in this example.
    MySrcConnCreateCbClass mySrcCreateCbObj = new MySrcConnCreateCbClass();
    MySrcConnDeleteCbClass mySrcDeleteCbObj = new MySrcConnDeleteCbClass();
    // Object passed as "cbArg" to myUmSrcEventCb, mySrcCreateCbObj, and mySrcDeleteCbObj.
    Object mySrcCbArg = null;
    // Create and initialize the connected source.
    LbmCtSrc myCtSrc = new LbmCtSrc();  // This just creates the object.
    myCtSrc.start(myCt, "myTopic", mySrcAttr, myUmSrcEventCb, mySrcCreateCbObj, mySrcDeleteCbObj, mySrcCbArg);

    for (int i = 0; i < 10; i++) {
      Thread.sleep(1000);
      System.out.println("Sending...");
      myCtSrc.getUmSrc().send("TestMsg".getBytes(), 7, LBM.MSG_FLUSH);
    }

    // Clean up.
    myCtSrc.stop();  // Important to stop the CT Source when finished.
    myCt.stop();  //  Important to stop the CT object when finished.
  }


  // Callback for when a CT receiver connects to this source.
  class MySrcConnCreateCbClass implements LbmCtSrcConnCreateCallback {
    @Override
    public Object onSrcConnCreate(LbmCtSrc ctSrc, LbmCtPeerInfo peerInfo, Object cbArg) {
      // Put your connect code here.
      try {
        System.out.println("onSrcConnCreate: rcv metadata=" +
            new String(peerInfo.getRcvMetadata().array(), StandardCharsets.ISO_8859_1));
      } catch (Exception e) { System.out.println("Exception: " + e.toString()); }

      // It is typical to return an application-specific, per-connection state object here.
      // It is passed to the connection delete callback.
      return null;
    }
  }


  // Callback for when a CT receiver disconnects from this source.
  class MySrcConnDeleteCbClass implements LbmCtSrcConnDeleteCallback {
    @Override
    public void onSrcConnDelete(LbmCtSrc ctSrc, LbmCtPeerInfo peerInfo, Object cbArg, Object connCbArg) {
      // Put your disconnect code here.  Note that connCbArg is the object returned by onSrcConnCreate.
      try {
        System.out.println("onSrcConnDelete: rcv metadata=" +
            new String(peerInfo.getRcvMetadata().array(), StandardCharsets.ISO_8859_1));
      } catch (Exception e) { System.out.println("Exception: " + e.toString()); }
    }
  }


  public static void main(String[] args) throws Exception {
    MinCtSrc application = new MinCtSrc();
    application.run(args);
  }
}
